cpu/saml21/include/periph_cpu.h

```diff
@@ -41,6 +41,27 @@ extern "C" {
 #define SAML21_PM_MODE_IDLE    (2)  /**< CPU sleeping, peripherals are active */
 /** @} */
 
+/**
+ * @name    Peripheral power mode requirements
+ * @{
+ */
+#define SAM0_GPIO_PM_BLOCK        SAML21_PM_MODE_BACKUP   /**< GPIO IRQs require STANDBY mode */
+#define SAM0_RTCRTT_PM_BLOCK      SAML21_PM_MODE_BACKUP   /**< RTC/TRR require STANDBY mode */
```

Comment:
I'm pretty sure the RTT/RTC is always on as it can be used to wake from BACKUP sleep

> The Backup Power Domain (PDBACKUP) is always on, except in the off sleep mode. It contains the 32KHz oscillator sources, the Supply Controller, the Reset Controller, the Real Time Counter, and the Power Manager itself.

Reply:
Yes, right. But we're loosing context. Callback functions are never called, then ;-)

Maybe we can work with two cases:
1. A callback function has been specified (i.e. nun-NULL) with the `rtt_set_overflow_cb()`, `rtt_set_arlarm()`, `rtc_set_alarm()`. -> Block BACKUP
2. No callback function he been specified -> Don't block a power state at all.

Would that fit your use case?

Reply:
Ah yes of course.
But doesn't that mean BACKUP mode should never be entered automatically by `pm_layered` as we always lose all state there - it can only be entered explicitly.

Reply:
But if BACKUP must always be blocked, why add code to the drivers to block it and not just block it by default?

Reply:
I don't want all those exceptions anymore. Applications shouldn't interact with `periph_pm` directly. It feels smart, but I've seen so many bugs because of violating this rule.

You could also enter BACKUP by turning off the last active peripheral (which could be the uart). If all work is done, the MCU goes into BACKUP.

Reply:
But I don't see how this is going to be any easier with this. E.g. you want to enter Deep Sleep and either wake up after a set time (let's say 1h) or via GPIO interrupt (you've configured an interrupt on a wakeup enabled GPIO).

You can't rely on `pm_layered` for this because Deep Sleep is blocked by the driver. Yes you won't get the callback, but you can still read out what caused the wake from Deep Sleep.

If Deep Sleep management still has to remain manual, I'm not convinced that adding complexity to the RTC and GPIO driver has any benefit when this code will in fact never be used.

Reply:
True, but in that scenario we could still do with one (default) blocker of BACKUP and then have an app shutdown routine that does `pm_unblock(0)`.

That would be easier to manage than having to tell the GPIO and RTC driver that both now have to unblock BACKUP.

If the only wake up sources for BACKUP *do* block BACKUP mode, we are back to manual `pm_set()` to enter Deep Sleep.

Reply:
> That would be easier to manage than having to tell the GPIO and RTC driver that both now have to unblock BACKUP.

I tell the GPIO driver to stop listening for IRQs. And the RTC driver is also turned off. Both have the methods to do so and it the *valid* thing to do here. Or am I the wrong way driver here?

The `pm_unblock(0)` is still error-prone if - for example - the `at86rf2xx` driver is waiting for a network packet. The IRQ GPIO is the only async part of this driver. Nothing else would stop the MCU to go down to BACKUP.

(Of course this driver must be fixed with `gpio_irq_disable()`/`gpio_irq_enable()` calls when it goes into sleep and wakes up again.)

Reply:
> If the only wake up sources for BACKUP do block BACKUP mode, we are back to manual pm_set() to enter Deep Sleep.

You are talking about GPIO IRQs? Or what are you thinking about.

Reply:
> I tell the GPIO driver to stop listening for IRQs. And the RTC driver is also turned off.

But then how are you going to wake up from Deep Sleep?
Maybe we need an additional API function for that. For the GPIO driver we could have `gpio_init_wake()` and for the RTC/RTT driver we could have `rtx_set_wakeup()`?

The former would certainly be cleaner than the current approach where if `gpio_init_int()` is called on a pin that is capable of waking the system from Deep Sleep, it will just configure it to do so. (The advantage of this approach was that it required no driver change, if you connect your accelerometer to a wake pin and set the threshold even, it would just wake you up from Deep Sleep without the need to configure anything).

Reply:
I see ... yeah, we need the RTC ;-)

The correct way to do it (code size optimization out of the way) is to block and unblock in the `rtx_set_alarm()` / `rtx_clear_alarm()` resp. `rtt_set_overflow_cb()` / `rtt_clear_overflow_cb()` method and have a dedicated `rtx_set_wakeup()` method, right?

Reply:
39bf4a257942e34789287f8fa11afe1e5054bd32 tries to demonstrate that. The RTC/RTT driver is kinda crowded ...

Reply:
ROM size for `tests/periph_rtc` on the `samr30-xpro`:

```
# Driver from current master copied in place
   text	   data	    bss	    dec	    hex	filename
  13836	    160	   2316	  16312	   3fb8	/home/jue/Projects/ssv/RIOT/tests/periph_rtc/bin/samr30-xpro/tests_periph_rtc.elf

# With SAM0_RTCRTT_PM_BLOCK undefined (i.e. no pm in this driver)
   text	   data	    bss	    dec	    hex	filename
  13836	    160	   2316	  16312	   3fb8	/home/jue/Projects/ssv/RIOT/tests/periph_rtc/bin/samr30-xpro/tests_periph_rtc.elf

# With SAM0_RTCRTT_PM_BLOCK defined
   text	   data	    bss	    dec	    hex	filename
  13928	    160	   2316	  16404	   4014	/home/jue/Projects/ssv/RIOT/tests/periph_rtc/bin/samr30-xpro/tests_periph_rtc.elf
```

Reply:
Regarding `gpio_init_int` vs. wake-up. I'd prefer to have a special `gpio_wake_enable()` / `gpio_wake_disable()` interface. It's a special feature that boards/cpus may provide. This separation would give users the option to opt-out. On top we can do proper power management ;-)

What do you think?